core/schema/module.go

```diff
@@ -316,176 +333,841 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
-	id, err := self.CurrentModule(ctx)
+func (s *moduleSchema) updateCodegenAndRuntime(ctx context.Context, mod *core.Module) (*core.Module, error) {
+	if mod.NameField == "" || mod.SDKConfig == "" {
+		return nil, fmt.Errorf("module cannot be generated without both name and sdk set")
+	}
+
+	// run codegen + get the runtime container
+	sdk, err := s.sdkForModule(ctx, mod.Query, mod.SDKConfig, mod.GeneratedSourceDirectory)
 	if err != nil {
-		return inst, err
+		return nil, fmt.Errorf("failed to get module sdk: %w", err)
 	}
-	return id.Load(ctx, s.dag)
+
+	sourceSubpath := mod.GeneratedSourceSubpath
+
+	mod.Runtime, err = sdk.Runtime(ctx, mod, mod.GeneratedSourceDirectory, sourceSubpath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get module runtime: %w", err)
+	}
+
+	generatedCode, err := sdk.Codegen(ctx, mod, mod.GeneratedSourceDirectory, sourceSubpath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate code: %w", err)
+	}
+
+	err = s.dag.Select(ctx, mod.GeneratedSourceDirectory, &mod.GeneratedSourceDirectory,
+		dagql.Selector{
+			Field: "withDirectory",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String("/")},
+				{Name: "directory", Value: dagql.NewID[*core.Directory](generatedCode.Code.ID())},
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to overlay generated code onto source directory: %w", err)
+	}
+
+	// update .gitattributes
+	gitAttrsPath := filepath.Join(sourceSubpath, ".gitattributes")
+	var gitAttrsContents []byte
+	gitAttrsFile, err := mod.GeneratedSourceDirectory.Self.File(ctx, gitAttrsPath)
+	if err == nil {
+		gitAttrsContents, err = gitAttrsFile.Contents(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get git attributes file contents: %w", err)
+		}
+		if !bytes.HasSuffix(gitAttrsContents, []byte("\n")) {
+			gitAttrsContents = append(gitAttrsContents, []byte("\n")...)
+		}
+	}
+	for _, fileName := range generatedCode.VCSGeneratedPaths {
+		if bytes.Contains(gitAttrsContents, []byte(fileName)) {
+			// already has some config for the file
+			continue
+		}
+		gitAttrsContents = append(gitAttrsContents,
+			[]byte(fmt.Sprintf("/%s linguist-generated=true\n", fileName))...,
+		)
+	}
+
+	err = s.dag.Select(ctx, mod.GeneratedSourceDirectory, &mod.GeneratedSourceDirectory,
+		dagql.Selector{
+			Field: "withNewFile",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(gitAttrsPath)},
+				{Name: "contents", Value: dagql.String(gitAttrsContents)},
+				{Name: "permissions", Value: dagql.Int(0600)},
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add git attributes file: %w", err)
+	}
+
+	return mod, nil
 }
 
-func (s *moduleSchema) currentFunctionCall(ctx context.Context, self *core.Query, _ struct{}) (*core.FunctionCall, error) {
-	return self.CurrentFunctionCall(ctx)
+type moduleSourceArgs struct {
+	// avoiding name "ref" due to that being a reserved word in some SDKs (e.g. Rust)
```

Comment:
dang, unfortunate

Reply:
Yeah, it should be a requirement for every SDK to have a list of reserved words and deal with them from day one like Python does, rather than having to adjust manually when they come up:

https://github.com/dagger/dagger/blob/f7d1805f812b4ff58a08611eef63d6bffa3e5cee/sdk/python/src/dagger/_codegen/generator.py#L416-L417